### PR TITLE
Mime command

### PR DESCRIPTION
**Pull request by kik4444**
# Description

Initial implementation of #12657.
Add a new dedicated series of commands for working with mime types without disk access.
- `mime` - simply a help command like `str`
- `mime list` - returns list of extensions for the given mime
- `mime guess` - returns mime type guesses for the given file path or extension. Input / Output is `string -> string` and `list<string> -> table<name: string, type: string>`

~~I have not yet figured out how to work with `PipelineData::ListStream` and until I do this will be a draft PR.~~

Everything shown in the examples works.

I am not too familiar with the internals of the Nushell codebase, so please let me know if I've made any mistakes or inefficiencies, such as with the spans.

What feedback do you have for the current state of these commands?

# User-Facing Changes

New commands for working with mime types. Here's a sample from the help pages:
```nu
Get known extensions for the "video/x-matroska" mime type
  > mime list "video/x-matroska"
  ╭───┬──────╮
  │ 0 │ mk3d │
  │ 1 │ mks  │
  │ 2 │ mkv  │
  ╰───┴──────╯

  Get all known video extensions
  > mime list "video/*"
```
```nu
Guess the MIME type from the path and return a string.
  > "video.mkv" | mime guess
  "video/x-matroska"

  Guess the MIME types from the paths and return a table.
  > ["video.mkv" "audio.mp3"] | mime guess
  ╭───┬───────────┬──────────────────╮
  │ # │   name    │       type       │
  ├───┼───────────┼──────────────────┤
  │ 0 │ video.mkv │ video/x-matroska │
  │ 1 │ audio.mp3 │ audio/mpeg       │
  ╰───┴───────────┴──────────────────╯

  Guess the MIME types from the extensions and return a table.
  > ["mkv" "mp3"] | mime guess -e
  ╭───┬──────┬──────────────────╮
  │ # │ name │       type       │
  ├───┼──────┼──────────────────┤
  │ 0 │ mkv  │ video/x-matroska │
  │ 1 │ mp3  │ audio/mpeg       │
  ╰───┴──────┴──────────────────╯

  Add a MIME type column to a table.
  > let input = glob -d 1 * | wrap filename; $input | merge ($input | get filename | mime guess | select type)
  ╭───┬──────────┬──────╮
  │ # │ filename │ type │
  ├───┼──────────┼──────┤
  │ 0 │ ...      │ ...  │
  ╰───┴──────────┴──────╯
```

# Tests + Formatting

~~`cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` currently fails because of an unused variable in the `PipelineData::ListStream` section.~~

# After Submitting

### CODE REVIEW

**Comment by sholderbach** [comment]
Should this start its life out as a plugin?

Open question for me around #12657 
- implementation cost seems low rn but is there a stability risk for nushell as the [crate doesn't guarantee that the mime-type assignment is stable](https://docs.rs/mime_guess/2.0.4/mime_guess/#note-mime-types-returned-are-not-stableguaranteed)?
- What is the right source of "truth" for mime type/extension mappings? A statically baked form in the binary, the system database like in `/usr/share/mime`?
(the same discussion could apply to whether `ls --mime-type` is "stabilizable" or should overload the `type` column)

**Comment by kik4444** [comment]
> Should this start its life out as a plugin?
> 
> Open question for me around #12657
> 
>     * implementation cost seems low rn but is there a stability risk for nushell as the [crate doesn't guarantee that the mime-type assignment is stable](https://docs.rs/mime_guess/2.0.4/mime_guess/#note-mime-types-returned-are-not-stableguaranteed)?
> 
>     * What is the right source of "truth" for mime type/extension mappings? A statically baked form in the binary, the system database like in `/usr/share/mime`?
>       (the same discussion could apply to whether `ls --mime-type` is "stabilizable" or should overload the `type` column)

I think the PR is mostly done now. As for your question, you're right that anything you can ask about this can also be asked about `ls`'s `--mime-type` flag.

Regarding where we get mime associations, I don't think the `mime_guess` crate supports anything besides its own hardcoded associations. Though does every platform supported by Nushell even have something like a `/usr/share/mime` database?

File: crates/nu-command/src/strings/mime_/guess.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    ) -> Result<PipelineData, ShellError> {
+        let use_extension: bool = call.has_flag(engine_state, stack, "extension")?;
+
+        let guess_function: fn(&str) -> mime_guess::MimeGuess = if use_extension {
+            mime_guess::from_ext
+        } else {
+            // HACK I don't know how to satisfy the compiler here without a closure, but I cannot return the function directly.
+            // If I do, I get an error that the types are different or that a value does not live long enough when the function is called.
+            |input| mime_guess::from_path(input)
+        };
```
Comment (sholderbach) [inline crates/nu-command/src/strings/mime_/guess.rs:111]: I guess that is due to the generic `P: AsRef<Path>` on `from_path` compared to the concrete `&str` parameter in `from_ext`?

File: crates/nu-command/src/strings/mime_/guess.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+                            Value::record(record!("name" => name, "type" => mime_type), span)
+                        }
+                        Err(err) => Value::error(
+                            ShellError::TypeMismatch {
+                                err_message: err.to_string(),
+                                span,
+                            },
+                            span,
+                        ),
```
Comment (sholderbach) [inline crates/nu-command/src/strings/mime_/guess.rs:145]: `to_string`ing the whole original error seems wrong. Either simply keep the `ShellError::CantConvert` from `Value::as_str` or do a proper `ShellError::OnlySupportsThisInputType`

**Review by sholderbach — COMMENTED** [review]
If it were just up to me, who favors predictability/long-term maintenance over some features, I would advocate for the removal of `ls --mime-type` and see the replacement in your commands shipped as a plugin, until there is a overwhelming need and proven reliability of the implementation.

But others may have different priorities and can resolve the stability/platform-support questions.

**Comment by kik4444** [comment]
Personally I'd prefer these commands to be builtin cause that makes them easier to use and discover, but I'm already making them for a *probably* niche use-case so ultimately I wouldn't mind if they were a plugin either. I'll leave that decision to the Nu higher-ups.

File: crates/nu-command/src/strings/mime_/guess.rs (hunk 1 above)
Comment (kik4444) [reply crates/nu-command/src/strings/mime_/guess.rs:111]: Yes, the genericness of `from_path` is not playing nice with the concreteness of `from_ext`. That's the reason for the error that the types mismatch. That's why I try to coerce it into the same type with a closure. Hopefully the compiler will optimize that away.

However, another error shows up later when I use the variable:  `guess_function` does not live long enough. I'm not sure why that one is caused. Isn't it supposed to be a cheaply `Copy`-able fn pointer?

**Comment by kik4444** [comment]
> removal of `ls --mime-type` and see the replacement in your commands shipped as a plugin

I must emphasize, however, that `ls --mime-type` does more than my commands. My commands are a glorified wrapper over `mime_guess`. One notable difference is that `ls` checks if a path is a directory and then returns `dir` on the `type` column. This `mime guess` command does not perform disk access, so passing a name to it without an extension will return `"unknown"`. And directories are all basically names without extensions.

I've already mentioned this behavior in the usage part of `mime guess`.

**Comment by fdncred** [comment]
There are a lot of commands that I could live with as plugins, this is one of them. I'm pretty sure that I wrote the `ls --mime-type` because someone asked for it, but for my usage, it is pretty niche. 

The thing that I always come back to is that nushell is a "batteries included" shell. If we remove all the batteries, it's not as well rounded and helpful but at what point are there enough batteries? I'm not sure.

**Comment by kik4444** [comment]
Okay, I'll rewrite this to be a plugin soon. I'll make this into a draft again until it's ready. If that's not okay you can go ahead and close this PR and I'll remake it with the changes.

**Comment by kik4444** [comment]
I'm going to need some help @fdncred. Since I want `mime guess` to work with streams, I need to provide `ctrlc` to `into_pipeline_data`, but how do you get `ctrlc` inside a plugin? I don't see any fields or methods on `engine: &nu_plugin::EngineInterface` that can help and I can't find any examples of this elsewhere in the code.

**Comment by fdncred** [comment]
I'm not sure ctrlc is exposed to plugins yet. Is it @devyn?

**Comment by kik4444** [comment]
Alright I'll push what I have so far anyway. It might make it easier to see what I'm trying to do.

File: crates/nu_plugin_mime/src/guess.rs [hunk 3]
```diff
@@ -124,15 +131,16 @@ impl Command for MimeGuess {
                     }
                 });
 
-                let ctrlc = engine_state.ctrlc.clone();
+                let ctrlc = compile_error!("can't figure out how to get ctrlc in plugin yet");
```
Comment (kik4444) [inline crates/nu_plugin_mime/src/guess.rs:134]: This is the part in question where I need ctrlc to be able to make this streamable.

**Comment by kik4444** [comment]
Hello, any news on this @fdncred?

**Comment by fdncred** [comment]
Which part? We decided it needs to be a plugin. I don't think we have resolved the ctrlc part for plugins.

**Comment by kik4444** [comment]
That's the part. Any news about ctrlc for plugins?

**Comment by devyn** [comment]
Hi @kik4444 

It's not necessary to pass ctrlc to make a stream for a plugin. The engine side adds it for you, and handles the messages to cancel the stream on the plugin side.

You can always just pass `None` inside a plugin and it will work fine.

**Comment by kik4444** [comment]
Hello @devyn,

Is there an issue in the way nushell passes `ListStream`s to plugins? Because now that I set ctrlc to `None` the plugin does indeed work, but whenver I pipe a `ListStream` to it I get an error even before the plugin's function is invoked.
```nu
> glob * | mime guess
Plugin `mime` error: Plugin failed to decode: missing field `span`
Error:   × Failed to receive response to plugin call from `mime`
   ╭─[entry #18:1:10]
 1 │ glob * | mime guess
   ·          ─────┬────
   ·               ╰── while waiting for this operation to complete
   ╰────
  help: try restarting the plugin with `plugin use 'mime'`
```

**Comment by devyn** [comment]
Hi @kik4444,

I believe your `nu` and `nu-plugin` versions are out of sync. It sounds like one of the two of them is on `main`, while the other is probably on 0.93.0

**Comment by kik4444** [comment]
You were right. Now it seems to work as far as I can tell.

**Comment by fdncred** [comment]
I think this should be in a separate repo and not in nushell. Then we can add it to the awesome-nu repo and people can use it.

**Comment by kik4444** [comment]
Okay, if that's the case you could go ahead close the PR and I'll eventually upload it to my profile

**Comment by fdncred** [comment]
Please ping me when you have it in your repo.

**Comment by kik4444** [comment]
@fdncred Ok, sorry for the delay. Here it is https://github.com/kik4444/nu_plugin_mime
I believe it's complete. I updated the dependencies and `repository` field in the Cargo.toml so I don't think there's anything left that I've missed.

**Comment by fdncred** [comment]
@kik4444 very cool! Thanks. Feel free to submit a PR to our awesome-nu repo to point to your plugin.

**Pull request closed without merging**